Create packages for bigquery and outline spi layer

mziccard opened the pull request:
First commit for bigquery module.
Main remarks on `BigqueryRpc`:
- In bigquery some list endpoints also return the number of total items (list tables and list table data). That might be a useful information to return. Do you think we should return a triple instead of a pair? Our `Page` interface however does not support such a property.
- list and get methods have explicit names (e.g. `listJobs`, `listTables`) as some share the same signature

aozarov commented on gcloud-java-bigquery/README.md:
```diff
@@ -0,0 +1,86 @@
+Data can be streamed into BigQuery at millions of rows per second to enable real-time analysis.
+With BigQuery you can easily deploy Petabyte-scale Databases.
+
+Be sure to activate the Google Cloud Resource Manager API on the Developer's Console to use Resource Manager from your project.
```
s/ Resource Manager AP/BigQuery API/ and s/Resource Manager/BigQuery/ ?

mziccard commented:
I almost forgot. Some of the list APIs return partial resources not containing all information (see [list tables](https://cloud.google.com/bigquery/docs/reference/v2/tables/list) for instance).

Such partial resources are mapped to dedicated model classes in the java API (see [ListTables](https://developers.google.com/resources/api-libraries/documentation/bigquery/v2/java/latest/com/google/api/services/bigquery/model/TableList.Tables.html) for an example).

I decided not to expose such classes in our spi layer but rather transform partial resources to a complete resource. This is open for discussion and I would like to hear you toughts.

mziccard commented:
I rebased to include into the branch the new `BaseServiceException` and fixed the comment on the README.

CLA's bot is complaining as I included other people's commit but all authors have already signed the CLA.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/BigqueryRpc.java:
```diff
@@ -0,0 +1,142 @@
+    PAGE_TOKEN("pageToken"),
+    START_INDEX("startIndex"),
+    STATE_FILTER("stateFilter"),
+    TIMEOUT("timeOut");
```
timeoutMs?

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/BigqueryRpc.java:
```diff
@@ -0,0 +1,142 @@
+public interface BigqueryRpc {
+
+  // These options are part of the Google Cloud BigQuery query parameters
+  enum Option {
```
Do we plan to always use projection=full (when [applicable](https://developers.google.com/resources/api-libraries/documentation/bigquery/v2/java/latest/com/google/api/services/bigquery/Bigquery.Jobs.List.html))?

mziccard replied:
Yes

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/BigqueryRpc.java:
```diff
@@ -0,0 +1,142 @@
+  TableDataInsertAllResponse insertAll(TableReference table, TableDataInsertAllRequest request,
+      Map<Option, ?> options) throws BigqueryException;
+
+  Tuple<String, Iterable<TableRow>> listTableData(String datasetId, String tableId,
```
Does not this miss some response data (such as [kind](https://developers.google.com/resources/api-libraries/documentation/bigquery/v2/java/latest/com/google/api/services/bigquery/model/TableDataList.html#getKind()), and [etag](https://developers.google.com/resources/api-libraries/documentation/bigquery/v2/java/latest/com/google/api/services/bigquery/model/TableDataList.html#getEtag()))?

mziccard replied:
Well `kind` is the type of response (`bigquery#tableDataList`) and it's usually there in the APIs but we never used it.

`etag` might be used to check the sanity of the response but since the list is paginated we are most likely going to use `Page` which does not expose any `etag` information.

aozarov commented:
Done with my pass over the PR.

Answers for open questions:

> In bigquery some list endpoints also return the number of total items (list tables and list table data). That might be a useful information to return. Do you think we should return a triple instead of a pair? Our Page interface however does not support such a property.

I am fine with leaving it out for now. We can always extend Page if we need to however I doubt such
information is that useful (as I suspect the count is not guaranteed to be the same count of the total
values in the iterated pages).

> Such partial resources are mapped to dedicated model classes in the java API (see ListTables for an example).
I decided not to expose such classes in our spi layer but rather transform partial resources to a complete resource. This is open for discussion and I would like to hear your thoughts.

I am fine with that and with documentation. If we supported "replace" operation I would have suggested
to add a flag "partial" but with only supporting "update"/patch I think we should be OK without it.

mziccard commented:
I fixed the comments and renamed Bigquery to BigQuery
